internal/log/hook.go

```diff
@@ -1,35 +1,147 @@
 package log
 
 import (
+	"bytes"
+	"reflect"
+	"time"
+
 	"github.com/Microsoft/hcsshim/internal/logfields"
+	"github.com/containerd/containerd/log"
 	"github.com/sirupsen/logrus"
 	"go.opencensus.io/trace"
 )
 
+const nullString = "null"
+
 // Hook serves to intercept and format `logrus.Entry`s before they are passed
 // to the ETW hook.
 //
 // The containerd shim discards the (formatted) logrus output, and outputs only via ETW.
 // The Linux GCS outputs logrus entries over stdout, which is consumed by the shim and
 // then re-output via the ETW hook.
-type Hook struct{}
+type Hook struct {
+	// EncodeAsJSON formats structs, maps, arrays, slices, and `bytes.Buffer` as JSON.
+	// Variables of `bytes.Buffer` will be converted to `[]byte`.
+	//
+	// Default is true.
+	EncodeAsJSON bool
+
+	// FormatTime specifies the format for `time.Time` variables.
+	// An empty string disabled formatting.
+	//
+	// Default is `"github.com/containerd/containerd/log".RFC3339NanoFixed`.
+	TimeFormat string
+
+	// AddSpanContext adds `logfields.TraceID` and `logfields.SpanID` fields to
+	// the entry from the span context stored in `(*entry).Context`, if it exists.
+	AddSpanContext bool
+
+	// Whether to encode errors or keep them as is.
+	EncodeError bool
+}
 
 var _ logrus.Hook = &Hook{}
 
 func NewHook() *Hook {
-	return &Hook{}
+	return &Hook{
+		EncodeAsJSON:   true,
+		TimeFormat:     log.RFC3339NanoFixed,
+		AddSpanContext: true,
+	}
 }
 
 func (h *Hook) Levels() []logrus.Level {
 	return logrus.AllLevels
 }
 
 func (h *Hook) Fire(e *logrus.Entry) (err error) {
+	// JSON encode, if necessary, then add span information
+	h.encode(e)
 	h.addSpanContext(e)
 
 	return nil
 }
 
+func (h *Hook) encode(e *logrus.Entry) {
+	d := e.Data
+
+	formatTime := len(h.TimeFormat) > 0
+	if !(h.EncodeAsJSON || formatTime) {
+		return
+	}
+
+	// todo: replace these with constraints.Integer, constraints.Float, etc with go1.18
+	for k, v := range d {
+		if !h.EncodeError {
+			if _, ok := v.(error); k == logrus.ErrorKey || ok {
+				continue
+			}
+		}
+
+		if formatTime {
+			switch vv := v.(type) {
+			case time.Time:
+				d[k] = vv.Format(h.TimeFormat)
+				continue
+			}
+		}
+
+		if !h.EncodeAsJSON {
+			continue
+		}
+
+		switch vv := v.(type) {
+		// built in types
+		case bool, string, error, uintptr,
+			int8, int16, int32, int64, int,
+			uint8, uint32, uint64, uint,
+			float32, float64:
+			continue
+
+		case time.Duration:
+			d[k] = vv.String()
```

Comment:
Might be nice to format this in uniform units (e.g. milliseconds), so it's easier to parse and work with. We previously put the span duration as a string duration and it has made some things difficult.